File: internal/models/notification/dingtalk.go
// Package notification 定义消息通知相关的数据模型
// 本文件包含钉钉相关的模型定义
package notification

import (
	"time"
)

// ==================== 钉钉模型 ====================

// DingtalkApp 钉钉应用模型
// 存储钉钉开放平台应用配置
type DingtalkApp struct {
	ID          uint      `gorm:"primarykey" json:"id"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	Name        string    `gorm:"size:100;not null" json:"name"`                         // 应用名称
	AppKey      string    `gorm:"column:app_key;size:100;not null" json:"app_key"`       // 钉钉 App Key
	AppSecret   string    `gorm:"column:app_secret;size:200;not null" json:"app_secret"` // 钉钉 App Secret
	AgentID     int64     `gorm:"column:agent_id" json:"agent_id"`                       // 钉钉 Agent ID
	Project     string    `gorm:"size:100" json:"project"`                               // 关联项目
	Status      string    `gorm:"size:20;default:'active';not null" json:"status"`       // 状态
	Description string    `gorm:"type:text" json:"description"`                          // 描述
	IsDefault   bool      `gorm:"column:is_default;default:false" json:"is_default"`     // 是否默认应用
	CreatedBy   *uint     `gorm:"column:created_by" json:"created_by"`
}

// TableName 指定表名
func (DingtalkApp) TableName() string {
	return "dingtalk_apps"
}

// DingtalkBot 钉钉机器人模型
// 存储钉钉 Webhook 机器人配置
type DingtalkBot struct {
	ID          uint      `gorm:"primarykey" json:"id"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	Name        string    `gorm:"size:100;not null" json:"name"`                   // 机器人名称
	WebhookURL  string    `gorm:"size:500;not null" json:"webhook_url"`            // Webhook URL
	Secret      string    `gorm:"size:100" json:"secret"`                          // 加签密钥
	Description string    `gorm:"type:text" json:"description"`                    // 描述
	Status      string    `gorm:"size:20;default:'active';not null" json:"status"` // 状态
	CreatedBy   *uint     `gorm:"index" json:"created_by"`
}

// TableName 指定表名
func (DingtalkBot) TableName() string {
	return "dingtalk_bots"
}

// DingtalkMessageLog 钉钉消息发送记录
// 记录所有通过钉钉发送的消息
type DingtalkMessageLog struct {
	ID        uint      `gorm:"primarykey" json:"id"`
	CreatedAt time.Time `json:"created_at"`
	MsgType   string    `gorm:"size:50;not null" json:"msg_type"`        // 消息类型
	Target    string    `gorm:"size:200" json:"target"`                  // 发送目标
	Content   string    `gorm:"type:text" json:"content"`                // 消息内容
	Title     string    `gorm:"size:200" json:"title"`                   // 标题
	Source    string    `gorm:"size:50" json:"source"`                   // 来源
	Status    string    `gorm:"size:20;default:'success'" json:"status"` // 状态
	ErrorMsg  string    `gorm:"type:text" json:"error_msg"`              // 错误信息
	AppID     uint      `gorm:"column:app_id" json:"app_id"`             // 使用的钉钉应用ID
}

// TableName 指定表名
func (DingtalkMessageLog) TableName() string {
	return "dingtalk_message_logs"
}

File: web/src/services/dingtalk.ts
import request from './api'
import type { ApiResponse } from '../types'

// 钉钉应用
export interface DingtalkApp {
  id?: number
  name: string
  app_key: string
  app_secret: string
  agent_id: number
  project: string
  description?: string
  status: string
  is_default: boolean
  created_at?: string
  updated_at?: string
}

// 钉钉机器人
export interface DingtalkBot {
  id?: number
  name: string
  webhook_url: string
  secret?: string
  description?: string
  status: string
  created_at?: string
  updated_at?: string
}

// 钉钉消息日志
export interface DingtalkMessageLog {
  id: number
  created_at: string
  msg_type: string
  target: string
  content: string
  title?: string
  source: string
  status: string
  error_msg?: string
  app_id?: number
}

// 钉钉用户
export interface DingtalkUser {
  userid: string
  name: string
  mobile?: string
  email?: string
  avatar?: string
}

// 发送工作通知请求
export interface SendMessageRequest {
  app_id?: number
  userid_list: string
  msg_type: 'text' | 'markdown'
  content: string
  title?: string
}

// 发送Webhook请求
export interface SendWebhookRequest {
  bot_id: number
  msg_type: 'text' | 'markdown'
  content: { content?: string; title?: string; text?: string }
  at_all?: boolean
  at_users?: string[]
}


// 应用绑定信息
export interface AppBindings {
  jenkins_instances: Array<{ id: number; name: string; url: string }>
  k8s_clusters: Array<{ id: number; name: string }>
}

// 钉钉应用管理 API
export const dingtalkAppApi = {
  list: (page = 1, pageSize = 100): Promise<ApiResponse<{ list: DingtalkApp[]; total: number }>> => {
    return request.get('/dingtalk/app', { params: { page, page_size: pageSize } })
  },

  get: (id: number): Promise<ApiResponse<DingtalkApp>> => {
    return request.get(`/dingtalk/app/${id}`)
  },

  create: (data: Partial<DingtalkApp>): Promise<ApiResponse<DingtalkApp>> => {
    return request.post('/dingtalk/app', data)
  },

  update: (id: number, data: Partial<DingtalkApp>): Promise<ApiResponse<DingtalkApp>> => {
    return request.put(`/dingtalk/app/${id}`, data)
  },

  delete: (id: number): Promise<ApiResponse> => {
    return request.delete(`/dingtalk/app/${id}`)
  },

  setDefault: (id: number): Promise<ApiResponse> => {
    return request.post(`/dingtalk/app/${id}/default`)
  },

  getBindings: (id: number): Promise<ApiResponse<AppBindings>> => {
    return request.get(`/dingtalk/app/${id}/bindings`)
  }
}

// 钉钉机器人管理 API
export const dingtalkBotApi = {
  list: (page = 1, pageSize = 100): Promise<ApiResponse<{ list: DingtalkBot[]; total: number }>> => {
    return request.get('/dingtalk/bot', { params: { page, page_size: pageSize } })
  },

  get: (id: number): Promise<ApiResponse<DingtalkBot>> => {
    return request.get(`/dingtalk/bot/${id}`)
  },

  create: (data: Partial<DingtalkBot>): Promise<ApiResponse<DingtalkBot>> => {
    return request.post('/dingtalk/bot', data)
  },

  update: (id: number, data: Partial<DingtalkBot>): Promise<ApiResponse<DingtalkBot>> => {
    return request.put(`/dingtalk/bot/${id}`, data)
  },

  delete: (id: number): Promise<ApiResponse> => {
    return request.delete(`/dingtalk/bot/${id}`)
  }
}

// 钉钉消息 API
export const dingtalkApi = {
  sendMessage: (data: SendMessageRequest): Promise<ApiResponse> => {
    return request.post('/dingtalk/send-message', data)
  },

  sendWebhook: (data: SendWebhookRequest): Promise<ApiResponse> => {
    return request.post('/dingtalk/send-webhook', data)
  },

  searchUser: (query: string, appId?: number): Promise<ApiResponse<DingtalkUser[]>> => {
    return request.post('/dingtalk/user/search', { query, app_id: appId })
  }
}

// 钉钉消息日志 API
export const dingtalkLogApi = {
  list: (page = 1, pageSize = 20, msgType = '', source = ''): Promise<ApiResponse<{ list: DingtalkMessageLog[]; total: number }>> => {
    return request.get('/dingtalk/logs', { params: { page, page_size: pageSize, msg_type: msgType, source } })
  }
}
